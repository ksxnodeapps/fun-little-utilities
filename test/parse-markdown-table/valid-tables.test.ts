import {
  createMarkdownObjectTable,
  createMarkdownArrayTable,
  createMarkdownArrayTableSync,
  createMarkdownObjectTableSync,
} from 'parse-markdown-table'
import { getAsyncArray } from './lib/async-array'

describe('full bound', () => {
  async function setup() {
    const text = `
      | id | name        | email                    |
      |----|-------------|--------------------------|
      |  1 | John Doe    | john-doe@gmail.com       |
      |  2 | Peter Smith | petersmith22@outlook.com |
      |  3 | Julia Jones | jjones778@gmail.com      |
    `

    const table = await createMarkdownObjectTable(text)
    const list = await getAsyncArray(table)
    return { text, table, list }
  }

  it('list matches snapshot', async () => {
    const { list } = await setup()
    expect(list).toMatchSnapshot()
  })
})

describe('left bound', () => {
  async function setup() {
    const text = `
      | id | name        | email
      |----|-------------|--------------------------
      |  1 | John Doe    | john-doe@gmail.com
      |  2 | Peter Smith | petersmith22@outlook.com
      |  3 | Julia Jones | jjones778@gmail.com
    `

    const table = await createMarkdownObjectTable(text)
    const list = await getAsyncArray(table)
    return { text, table, list }
  }

  it('list matches snapshot', async () => {
    const { list } = await setup()
    expect(list).toMatchSnapshot()
  })
})

describe('right bound', () => {
  async function setup() {
    const text = `
       id | name        | email                    |
      ----|-------------|--------------------------|
        1 | John Doe    | john-doe@gmail.com       |
        2 | Peter Smith | petersmith22@outlook.com |
        3 | Julia Jones | jjones778@gmail.com      |
    `

    const table = await createMarkdownObjectTable(text)
    const list = await getAsyncArray(table)
    return { text, table, list }
  }

  it('list matches snapshot', async () => {
    const { list } = await setup()
    expect(list).toMatchSnapshot()
  })
})

describe('no bound', () => {
  async function setup() {
    const text = `
       id | name        | email
      ----|-------------|--------------------------
        1 | John Doe    | john-doe@gmail.com
        2 | Peter Smith | petersmith22@outlook.com
        3 | Julia Jones | jjones778@gmail.com
    `

    const table = await createMarkdownObjectTable(text)
    const list = await getAsyncArray(table)
    return { text, table, list }
  }

  it('list matches snapshot', async () => {
    const { list } = await setup()
    expect(list).toMatchSnapshot()
  })
})

describe('unknown columns', () => {
  async function setup() {
    const text = `
      | id | name        | email                    |
      |----|-------------|--------------------------|
      |  1 | John Doe    | john-doe@gmail.com       |  123 | male | true |
      |  2 | Peter Smith | petersmith22@outlook.com |
      |  3 | Julia Jones | jjones778@gmail.com      |  12  | female |
    `

    const table = await createMarkdownObjectTable(text)
    const list = await getAsyncArray(table)
    return { text, table, list }
  }

  it('list matches snapshot', async () => {
    const { list } = await setup()
    expect(list).toMatchSnapshot()
  })
})

async function* getChunks(text: string, splitter: string) {
  const [first, ...rest] = text.split(splitter)
  yield first
  yield* rest.map(chunk => splitter + chunk)
}

describe('iterate by lines', () => {
  async function setup() {
    const text = `
      | id | name        | email                    |
      |----|-------------|--------------------------|
      |  1 | John Doe    | john-doe@gmail.com       |
      |  2 | Peter Smith | petersmith22@outlook.com |
      |  3 | Julia Jones | jjones778@gmail.com      |
    `

    const table = await createMarkdownObjectTable(getChunks(text, '\n'))
    const list = await getAsyncArray(table)
    return { text, table, list }
  }

  it('list matches snapshot', async () => {
    const { list } = await setup()
    expect(list).toMatchSnapshot()
  })
})

describe('iterate by chunks', () => {
  async function setup() {
    const text = `
      | id | name        | email                    |
      |----|-------------|--------------------------|
      |  1 | John Doe    | john-doe@gmail.com       |
      |  2 | Peter Smith | petersmith22@outlook.com |
      |  3 | Julia Jones | jjones778@gmail.com      |
    `

    const table = await createMarkdownObjectTable(getChunks(text, '|'))
    const list = await getAsyncArray(table)
    return { text, table, list }
  }

  it('list matches snapshot', async () => {
    const { list } = await setup()
    expect(list).toMatchSnapshot()
  })
})

describe('createMarkdownArrayTable', () => {
  async function setup() {
    const text = `
      | id | name        | email                    |
      |----|-------------|--------------------------|
      |  1 | John Doe    | john-doe@gmail.com       |
      |  2 | Peter Smith | petersmith22@outlook.com |
      |  3 | Julia Jones | jjones778@gmail.com      |
    `

    const table = await createMarkdownArrayTable(text)
    const headers = table.headers
    const rows = await getAsyncArray(table.rows)
    return { text, table, headers, rows }
  }

  it('headers matches snapshot', async () => {
    const { headers } = await setup()
    expect(headers).toMatchSnapshot()
  })

  it('headers matches snapshot', async () => {
    const { rows } = await setup()
    expect(rows).toMatchSnapshot()
  })
})

describe('createMarkdownArrayTableSync', () => {
  function setup() {
    const text = `
      | id | name        | email                    |
      |----|-------------|--------------------------|
      |  1 | John Doe    | john-doe@gmail.com       |
      |  2 | Peter Smith | petersmith22@outlook.com |
      |  3 | Julia Jones | jjones778@gmail.com      |
    `

    const table = createMarkdownArrayTableSync(text)
    const headers = table.headers
    const rows = Array.from(table.rows)
    return { text, table, headers, rows }
  }

  it('headers matches snapshot', () => {
    const { headers } = setup()
    expect(headers).toMatchSnapshot()
  })

  it('headers matches snapshot', () => {
    const { rows } = setup()
    expect(rows).toMatchSnapshot()
  })
})

describe('createMarkdownObjectTableSync', () => {
  function setup() {
    const text = `
      | id | name        | email                    |
      |----|-------------|--------------------------|
      |  1 | John Doe    | john-doe@gmail.com       |
      |  2 | Peter Smith | petersmith22@outlook.com |
      |  3 | Julia Jones | jjones778@gmail.com      |
    `

    return Array.from(createMarkdownObjectTableSync(text))
  }

  it('list matches snapshot', () => {
    const list = setup()
    expect(list).toMatchSnapshot()
  })
})

describe('spaces in horizontal line', () => {
  async function setup() {
    const text = `
    | id | name        | email                    |
    | -- | ----------- |------------------------- |
    |  1 | John Doe    | john-doe@gmail.com       |
    |  2 | Peter Smith | petersmith22@outlook.com |
    |  3 | Julia Jones | jjones778@gmail.com      |
  `

    const table = await createMarkdownObjectTable(getChunks(text, '|'))
    const list = await getAsyncArray(table)
    return { text, table, list }
  }

  it('list matches snapshot', async () => {
    const { list } = await setup()
    expect(list).toMatchSnapshot()
  })
})
