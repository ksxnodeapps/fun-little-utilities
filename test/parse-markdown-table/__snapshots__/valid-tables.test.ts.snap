// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMarkdownArrayTable headers matches snapshot 1`] = `
Array [
  "id",
  "name",
  "email",
]
`;

exports[`createMarkdownArrayTable headers matches snapshot 2`] = `
Array [
  Array [
    "1",
    "John Doe",
    "john-doe@gmail.com",
  ],
  Array [
    "2",
    "Peter Smith",
    "petersmith22@outlook.com",
  ],
  Array [
    "3",
    "Julia Jones",
    "jjones778@gmail.com",
  ],
]
`;

exports[`createMarkdownArrayTableSync headers matches snapshot 1`] = `
Array [
  "id",
  "name",
  "email",
]
`;

exports[`createMarkdownArrayTableSync headers matches snapshot 2`] = `
Array [
  Array [
    "1",
    "John Doe",
    "john-doe@gmail.com",
  ],
  Array [
    "2",
    "Peter Smith",
    "petersmith22@outlook.com",
  ],
  Array [
    "3",
    "Julia Jones",
    "jjones778@gmail.com",
  ],
]
`;

exports[`createMarkdownObjectTableSync list matches snapshot 1`] = `
Array [
  Object {
    "email": "john-doe@gmail.com",
    "id": "1",
    "name": "John Doe",
  },
  Object {
    "email": "petersmith22@outlook.com",
    "id": "2",
    "name": "Peter Smith",
  },
  Object {
    "email": "jjones778@gmail.com",
    "id": "3",
    "name": "Julia Jones",
  },
]
`;

exports[`full bound list matches snapshot 1`] = `
Array [
  Object {
    "email": "john-doe@gmail.com",
    "id": "1",
    "name": "John Doe",
  },
  Object {
    "email": "petersmith22@outlook.com",
    "id": "2",
    "name": "Peter Smith",
  },
  Object {
    "email": "jjones778@gmail.com",
    "id": "3",
    "name": "Julia Jones",
  },
]
`;

exports[`iterate by chunks list matches snapshot 1`] = `
Array [
  Object {
    "email": "john-doe@gmail.com",
    "id": "1",
    "name": "John Doe",
  },
  Object {
    "email": "petersmith22@outlook.com",
    "id": "2",
    "name": "Peter Smith",
  },
  Object {
    "email": "jjones778@gmail.com",
    "id": "3",
    "name": "Julia Jones",
  },
]
`;

exports[`iterate by lines list matches snapshot 1`] = `
Array [
  Object {
    "email": "john-doe@gmail.com",
    "id": "1",
    "name": "John Doe",
  },
  Object {
    "email": "petersmith22@outlook.com",
    "id": "2",
    "name": "Peter Smith",
  },
  Object {
    "email": "jjones778@gmail.com",
    "id": "3",
    "name": "Julia Jones",
  },
]
`;

exports[`left bound list matches snapshot 1`] = `
Array [
  Object {
    "email": "john-doe@gmail.com",
    "id": "1",
    "name": "John Doe",
  },
  Object {
    "email": "petersmith22@outlook.com",
    "id": "2",
    "name": "Peter Smith",
  },
  Object {
    "email": "jjones778@gmail.com",
    "id": "3",
    "name": "Julia Jones",
  },
]
`;

exports[`no bound list matches snapshot 1`] = `
Array [
  Object {
    "email": "john-doe@gmail.com",
    "id": "1",
    "name": "John Doe",
  },
  Object {
    "email": "petersmith22@outlook.com",
    "id": "2",
    "name": "Peter Smith",
  },
  Object {
    "email": "jjones778@gmail.com",
    "id": "3",
    "name": "Julia Jones",
  },
]
`;

exports[`right bound list matches snapshot 1`] = `
Array [
  Object {
    "email": "john-doe@gmail.com",
    "id": "1",
    "name": "John Doe",
  },
  Object {
    "email": "petersmith22@outlook.com",
    "id": "2",
    "name": "Peter Smith",
  },
  Object {
    "email": "jjones778@gmail.com",
    "id": "3",
    "name": "Julia Jones",
  },
]
`;

exports[`unknown columns list matches snapshot 1`] = `
Array [
  Object {
    "email": "john-doe@gmail.com",
    "id": "1",
    "name": "John Doe",
    Symbol(Unknown Column #3): "123",
    Symbol(Unknown Column #4): "male",
    Symbol(Unknown Column #5): "true",
  },
  Object {
    "email": "petersmith22@outlook.com",
    "id": "2",
    "name": "Peter Smith",
  },
  Object {
    "email": "jjones778@gmail.com",
    "id": "3",
    "name": "Julia Jones",
    Symbol(Unknown Column #3): "12",
    Symbol(Unknown Column #4): "female",
  },
]
`;
